.github/actions/matrix-fan-out/action.yml

```diff
@@ -0,0 +1,33 @@
+name: "Matrix Fan-out"
+description: "Uploads a JSON file as an artifact with a unique name."
+
+inputs:
+  path:
+    description: "The directory to upload."
+    default: ./matrix-fan-out
+  retention-days:
+    description: "Optional. Defaults to repository settings."
+    default: null
```

Comment:
The repository settings are for the maximum limit for both logs and artifacts which is 90 days by default. Is there any reason a user would want their artifacts created via this action at any point outside of this workflow? I feel it should default to a lower value with the option to set it higher as the rarer case is wanting to have the retention set to a long period.

Reply:
Also, the default value is supposed to be a string according to GitHub: https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#inputsinput_iddefault. After taking a quick look at the implementation of actions/upload-artifact@v4, "null" (GitHub actions would turn this into a string) would be considered invalid as it is not a number nor is falsy. (ref https://github.com/actions/upload-artifact/blob/4c0ff1c489dca52fedb26375d7d8fe7bd9233f19/src/upload/input-helper.ts#L33C3-L39C4)

Reply:
If we want to keep the default as a falsy value (as interpreted by the upload action), I recommend using an empty string.

Reply:
> Also, the default value is supposed to be a string according to GitHub: https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#inputsinput_iddefault. After taking a quick look at the implementation of actions/upload-artifact@v4, "null" (GitHub actions would turn this into a string) would be considered invalid as it is not a number nor is falsy

You can know this is incorrect by seeing the successful selftest -- the null default is used, and there's no error about `"null" is not a valid number`.

Reply:
I understand that the current setup with null as the default value for retention-days seems to be working as expected. However, it's important to note that this goes against the intended schema for GitHub Actions, where all input values are treated as strings. While null is being interpreted in a way that doesn't cause issues right now, this behavior might change in future updates to GitHub Actions, potentially breaking our action.

It is also worth mentioning that my local GitHub actions linter is currently flagging this as an issue.

With that said, it is ultimately your decision to make.

Reply:
@IanWoodard  it does not seem to me that the github actions spec mentions anything about null values https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#inputsinput_iddefault

Reply:
@fpacifici You're correct that it doesn't explicitly mention how null values are handled. However, it does state that all input values are treated as strings, which a null literal is not one of. My concern is that while this setup works today, there's no guarantee it will work tomorrow. The behavior of null is not defined in the GitHub Actions documentation, which puts us in undefined territory. This could potentially lead to unexpected behavior or even break our action in future updates to GitHub Actions.